### PR TITLE
Allow to sort sidebar cameos

### PR DESCRIPTION
**Pull request by Thrifinesma**
[TechnoType]>CameoPriority=
Issue #66

### CODE REVIEW

File: src/Ext/TechnoType/Body.cpp [hunk 1]
```diff
@@ -99,6 +99,7 @@ void TechnoTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 	this->Promote_IncludeSpawns.Read(exINI, pSection, "Promote.IncludeSpawns");
 	this->ImmuneToCrit.Read(exINI, pSection, "ImmuneToCrit");
 	this->MultiMindControl_ReleaseVictim.Read(exINI, pSection, "MultiMindControl.ReleaseVictim");
+    this->CameoPriority.Read(exINI, pSection, "CameoPriority");
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.cpp:102]: Spaces instead of tabs here and everywhere below.
Also perhaps it would be better to have this ins `artmd.ini`?

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: src/Ext/TechnoType/Body.cpp (hunk 1 above)
Comment (Thrifinesma) [reply src/Ext/TechnoType/Body.cpp:102]: All sorting rules are in `rulesmd.ini`, so I don't suggestion the new one in `artmd.ini`

**Comment by FS-21** [comment]
For the tests I added CameoPriority=<integer value> to some few artillery units that I already know that in Unholy Alliance game mode (the one you can build everything of any faction) and 1 aircraft, another one that is a transport vehicle "considered aircraft", 2 naval units, 3 defenses of different factions and 2 Super Weapons.

I modified a testing UnitDelivery SW I had for deploy all the factories and structures that unlocks new units/structures.

What I saw after deploying the debug SW:
- In SW/defenses tab Super Weapons can't be ordered.
- In SW/defenses tab Structures Can't be placed before the Super Weapons.
- In SW/defenses tab the buildable Wall of the player's faction is placed before the first structures with "CameoPriority" tag. The wall didn't have any "CameoPriority" tag and you can't move the cameo adding a low value with "CameoPriority", looks hardcoded like Super Weapons.
- In SW/defenses tab **(ignoring all I said above)** The cameos with "CameoPriority" are ordered BEFORE the unordered cameos.
- In SW/defenses tab Structures with the same CameoPriority value are ordered by Tech Level (ordered from lower levels to highers levels).
- In vehicles tab ground vehicles are hardcoded to be placed first, then the aircrafts (and vehicles considered aircrafts) and finally naval units. In these 3 hardcoded groups the "CameoPriority" tag works as expected.

**Comment by FS-21** [comment]
Tested the newer build, I noticed:
- In units tab now any icon can be sorted ignoring the original 3 hardcoded groups (Ground units, Aircrafts & Naval).
- The (Player's) Wall (that depends of which House you selected) is still hardcoded to be the first icon after the Super Weapons.

File: src/Misc/Hooks.UI.cpp [hunk 2]
```diff
@@ -99,4 +101,33 @@ DEFINE_HOOK(0x715A4D, Replace_XXICON_With_New, 0x7)         //TechnoTypeClass::R
...
+		enum { rTrue = 0x6A8692, rFalse = 0x6A86A0 };
+
+		if (leftPriority > rightPriority)
+			return rTrue;
+		else if (rightPriority > leftPriority)
+			return rFalse;
+	}
+
+	// cmp esi, 1Fh
+	// jz short loc_6A8477
```
Comment (Metadorius) [inline src/Misc/Hooks.UI.cpp:130]:
```suggestion
	// Restore overridden instructions
```

File: src/Ext/TechnoType/Hooks.cpp [hunk 3]
```diff
@@ -99,4 +98,28 @@ DEFINE_HOOK(0x6B7282, SpawnManagerClass_AI_PromoteSpawns, 0x5)
...
+			return rTrue;
+		else if (rightPriority > leftPriority)
+			return rFalse;
+	}
+	// cmp esi, 28h
+	// jz short loc_6A855E
+	GET(AbstractType, rtti1, ESI);
+	return rtti1 == AbstractType::UnitType ? 0x6A855E : 0x6A853F;
+}
+*/
```
Comment (Metadorius) [inline src/Ext/TechnoType/Hooks.cpp:125]:
```suggestion
```

File: src/Ext/SWType/Hooks.cpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	auto leftPriority = pLeftExt->CameoPriority;
+	auto rightPriority = pRightExt->CameoPriority;
+	enum { rTrue = 0x6A8692, rFalse = 0x6A86A0 };
+	if (leftPriority > rightPriority)
+		return rTrue;
+	else if (rightPriority > leftPriority)
+		return rFalse;
+	return 0;
+}
+*/
```
Comment (Metadorius) [inline src/Ext/SWType/Hooks.cpp:18]: Not sure if there's any point in keeping this file at all.

**Review by Metadorius — CHANGES_REQUESTED** [review]
The code quality is OK, just a bit of changes and it can be merged.

**Pull request merged**